src/lib.rs

```diff
@@ -217,6 +217,28 @@ where
     }
 }
 
+impl<T, U> Array<MaybeUninit<T>, U>
+where
+    U: ArraySize,
+{
+    /// Create an uninitialized array of [`MaybeUninit`]s for the given type.
+    pub const fn uninit() -> Self {
+        // SAFETY: an array of `MaybeUninit`s is always valid.
+        unsafe { MaybeUninit::uninit().assume_init() }
+    }
```

Comment:
clippy doesn't like this:

```
error: this call for this type may be undefined behavior
   --> src/lib.rs:227:18
    |
227 |         unsafe { MaybeUninit::uninit().assume_init() }
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#uninit_assumed_init
    = note: `#[deny(clippy::uninit_assumed_init)]` on by default
```

I think it's a false positive... we're just assuming a type which wraps an array of `MaybeUninit`s is "initialized", where `MaybeUninit`'s idea of being "initialized" is that it may be uninitialized.

Reply:
Holy cow... No, it's not a false positive. It's a very serious and dangerous mistake, which was discussed numerous times. Even if all bit patterns are valid, you can not call `assume_init` on memory which was not properly initialized. You can not even create references to it (though it's a subject of discussion and ma change in future).

Reply:
The memory in question here is `[MaybeUninit<T>; N]`

Reply:
The [method docs](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.assume_init-1) are really explicit about that.

Reply:
Ideally we could use `Array(MaybeUninit::unit_array())` here but it's unstable, and written in a very similar manner:

```rust
    pub const fn uninit_array<const N: usize>() -> [Self; N] {
        // SAFETY: An uninitialized `[MaybeUninit<_>; LEN]` is valid.
        unsafe { MaybeUninit::<[MaybeUninit<T>; N]>::uninit().assume_init() }
    }
```

Reply:
@newpavlov yes, but isn't `MaybeUninit` itself the only valid exception?

Reply:
I was trying to make something that could be called in a `const` context but unfortunately could not similarly get `assume_init` to work in a const context, even when constraining `U: ArraySize<ArrayType = [T; N]>`.

If that doesn't work, I'm not sure how `ArrayOps` would help here, especially for a constructor.

Reply:
Yeah, trait method will not help here. Plus you will not be able to make it `const`.